File: packages/Teams/js/src/actions/sendMessageToTeamsChannel.ts
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

import {
  BoolExpression,
  BoolExpressionConverter,
  Expression,
  StringExpression,
  StringExpressionConverter,
} from 'adaptive-expressions';
import {
  Activity,
  Channels,
  CloudAdapterBase,
  ConversationReference,
  TeamsInfo,
} from 'botbuilder';
import {
  Converter,
  ConverterFactory,
  Dialog,
  DialogConfiguration,
  DialogContext,
  DialogStateManager,
  DialogTurnResult,
  TemplateInterface,
} from 'botbuilder-dialogs';
import {
  AuthenticationConstants,
  ClaimsIdentity,
  ConnectorClient,
  JwtTokenValidation,
} from 'botframework-connector';
import { ActivityTemplateConverter } from 'botbuilder-dialogs-adaptive/lib/converters';
import {
  getValue,
  testAdapterHasCreateConnectorClientMethod,
} from './actionHelpers';

export interface SendMessageToTeamsChannelConfiguration
  extends DialogConfiguration {
  disabled?: boolean | string | BoolExpression;
  property?: string | Expression | StringExpression;
  conversationReferenceProperty?: string | Expression | StringExpression;
  activityIdProperty?: string | Expression | StringExpression;
  teamsChannelId?: string | Expression | StringExpression;
  activity?: TemplateInterface<Activity, DialogStateManager>;
}

/**
 * Calls TeamsInfo.GetTeamDetails and sets the result to a memory property.
 */
export class SendMessageToTeamsChannel
  extends Dialog
  implements SendMessageToTeamsChannelConfiguration {
  /**
   * Class identifier.
   */
  static $kind = 'Teams.SendMessageToTeamsChannel';

  /**
   * Gets or sets an optional expression which if is true will disable this action.
   */
  public disabled?: BoolExpression;

  /**
   * Gets or sets property path to put the newly created activity's Conversation Reference.
   * This can be used to later send messages to this same conversation.
   */
  public conversationReferenceProperty?: StringExpression;

  /**
   *  Gets or sets property path to put the id of the activity sent.
   */
  public activityIdProperty?: StringExpression;

  /**
   * Gets or sets the expression to get the value to use for team id.
   *
   * @default
   * =turn.activity.channelData.channel.id
   */
  public teamsChannelId: StringExpression = new StringExpression(
    '=turn.activity.channelData.channel.id'
  );

  /**
   * Gets or sets template for the activity expression containing the activity to send.
   */
  public activity?: TemplateInterface<Activity, DialogStateManager>;

  public getConverter(
    property: keyof SendMessageToTeamsChannelConfiguration
  ): Converter | ConverterFactory {
    switch (property) {
      case 'disabled':
        return new BoolExpressionConverter();
      case 'conversationReferenceProperty':
      case 'activityIdProperty':
      case 'teamsChannelId':
        return new StringExpressionConverter();
      case 'activity':
        return new ActivityTemplateConverter();
      default:
        return super.getConverter(property);
    }
  }

  /**
   * Called when the dialog is started and pushed onto the dialog stack.
   *
   * @param {DialogContext} dc The [DialogContext](xref:botbuilder-dialogs.DialogContext) for the current turn of conversation.
   * @param {object} _options Optional, initial information to pass to the dialog.
   * @returns {Promise<DialogTurnResult>} A promise representing the asynchronous operation.
   */
  public async beginDialog(
    dc: DialogContext,
    _options?: Record<string, unknown>
  ): Promise<DialogTurnResult> {
    if (this.disabled?.getValue(dc.state)) {
      return dc.endDialog();
    }

    if (dc.context.activity.channelId !== Channels.Msteams) {
      throw new Error(
        `${SendMessageToTeamsChannel.$kind} works only on the Teams channel.`
      );
    }

    const activity = await this.activity?.bind(dc, dc.state);

    if (!activity) {
      throw new Error(
        `Missing Activity in ${SendMessageToTeamsChannel.$kind}.`
      );
    }

    const teamsChannelId =
      getValue(dc, this.teamsChannelId) ??
      dc.context?.activity.channelData?.channel?.id;

    let result: [ConversationReference, string];

    // Check for legacy adapter
    if (testAdapterHasCreateConnectorClientMethod(dc.context.adapter)) {
      const credentials = dc.context.turnState.get<ConnectorClient>(
        dc.context.adapter.ConnectorClientKey
      )?.credentials;
      if (!credentials) {
        throw new Error(
          'Missing credentials as MicrosoftAppCredentials in ConnectorClient from TurnState'
        );
      }

      result = await TeamsInfo.sendMessageToTeamsChannel(
        dc.context,
        activity,
        teamsChannelId
      );
    } else if (dc.context.adapter instanceof CloudAdapterBase) {
      const botIdentity = dc.context.turnState.get<ClaimsIdentity>(
        dc.context.adapter.BotIdentityKey
      );

      let appId: string | undefined;
      if (botIdentity) {
        // 'version' is sometimes empty, which will result in no id returned from GetAppIdFromClaims.
        appId = JwtTokenValidation.getAppIdFromClaims(botIdentity.claims);

        if (!appId) {
          appId = botIdentity.claims.find(
            (claim) => claim.type === AuthenticationConstants.AudienceClaim
          )?.value;
        }

        if (!appId) {
          throw new Error('Missing AppIdClaim in ClaimsIdentity');
        }
      } else {
        throw new Error(
          'Missing dc.context.adapter.BotIdentityKey in TurnContext TurnState'
        );
      }

      result = await TeamsInfo.sendMessageToTeamsChannel(
        dc.context,
        activity,
        teamsChannelId,
        appId
      );
    } else {
      throw new Error('The adapter does not support SendMessageToTeamsChannel');
    }

    if (this.conversationReferenceProperty != null) {
      dc.state.setValue(
        this.conversationReferenceProperty?.getValue(dc.state),
        result[0]
      );
    }

    if (this.activityIdProperty != null) {
      dc.state.setValue(this.activityIdProperty?.getValue(dc.state), result[1]);
    }

    return dc.endDialog(result);
  }

  /**
   * Builds the compute Id for the dialog.
   *
   * @returns {string} A string representing the compute Id.
   */
  protected onComputeId(): string {
    return `SendMessageToTeamsChannel[\
            ${this.teamsChannelId?.toString() ?? ''},\
            ${this.activityIdProperty?.toString() ?? ''},\
            ${this.conversationReferenceProperty?.toString() ?? ''}\
        ]`;
  }
}

File: packages/Teams/dotnet/Actions/SendMessageToTeamsChannel.cs
// Licensed under the MIT License.
// Copyright (c) Microsoft Corporation. All rights reserved.

using System;
using System.Linq;
using System.Runtime.CompilerServices;
using System.Security.Claims;
using System.Threading;
using System.Threading.Tasks;
using AdaptiveExpressions.Properties;
using Microsoft.Bot.Builder;
using Microsoft.Bot.Builder.Dialogs;
using Microsoft.Bot.Builder.Integration.AspNet.Core;
using Microsoft.Bot.Builder.Teams;
using Microsoft.Bot.Connector;
using Microsoft.Bot.Connector.Authentication;
using Microsoft.Bot.Schema;
using Newtonsoft.Json;

namespace Microsoft.Bot.Components.Teams.Actions
{
    /// <summary>
    /// Calls TeamsInfo.SendMessageToTeamsChannel and sets the result to a memory property.
    /// </summary>
    public class SendMessageToTeamsChannel : Dialog
    {
        /// <summary>
        /// Class identifier.
        /// </summary>
        [JsonProperty("$kind")]
        public const string Kind = "Teams.SendMessageToTeamsChannel";

        /// <summary>
        /// Initializes a new instance of the <see cref="SendMessageToTeamsChannel"/> class.
        /// </summary>
        /// <param name="callerPath">Optional, source file full path.</param>
        /// <param name="callerLine">Optional, line number in source file.</param>
        [JsonConstructor]
        public SendMessageToTeamsChannel([CallerFilePath] string callerPath = "", [CallerLineNumber] int callerLine = 0)
            : base()
        {
            RegisterSourceLocation(callerPath, callerLine);
        }

        /// <summary>
        /// Gets or sets an optional expression which if is true will disable this action.
        /// </summary>
        /// <example>
        /// "user.age > 18".
        /// </example>
        /// <value>
        /// A boolean expression. 
        /// </value>
        [JsonProperty("disabled")]
        public BoolExpression Disabled { get; set; } 

        /// <summary>
        /// Gets or sets property path to put the newly created activity's Conversation Reference.
        /// This can be used to later send messages to this same conversation.
        /// </summary>
        /// <value>
        /// Property path to put the newly created activity's conversation reference.
        /// </value>
        [JsonProperty("conversationReferenceProperty")]
        public StringExpression ConversationReferenceProperty { get; set; }

        /// <summary>
        /// Gets or sets property path to put the id of the activity sent.
        /// </summary>
        /// <value>
        /// Property path to put the id of the activity sent.
        /// </value>
        [JsonProperty("activityIdProperty")]
        public StringExpression ActivityIdProperty { get; set; }

        /// <summary>
        /// Gets or sets the expression to get the value to use for the teams channel id where the message should be sent. Default is turn.activity.channelData.channel.id.
        /// </summary>
        /// <value>
        /// The expression to get the value to use for teams channel id.
        /// </value>
        [JsonProperty("teamsChannelId")]
        public StringExpression TeamsChannelId { get; set; } = "=turn.activity.channelData.channel.id";

        /// <summary>
        /// Gets or sets template for the activity expression containing the activity to send.
        /// </summary>
        /// <value>
        /// Template for the activity.
        /// </value>
        [JsonProperty("activity")]
        public ITemplate<Activity> Activity { get; set; }

        /// <inheritdoc/>
        public override async Task<DialogTurnResult> BeginDialogAsync(DialogContext dc, object options = null, CancellationToken cancellationToken = default(CancellationToken))
        {
            if (options is CancellationToken)
            {
                throw new ArgumentException($"{nameof(options)} cannot be a cancellation token");
            }
            
            if (Disabled != null && Disabled.GetValue(dc.State))
            {
                return await dc.EndDialogAsync(cancellationToken: cancellationToken).ConfigureAwait(false);
            }

            if (dc.Context.Activity.ChannelId != Channels.Msteams)
            {
                throw new InvalidOperationException($"{Kind} works only on the Teams channel.");
            }

            IActivity activity = null;
            if (Activity != null)
            {
                activity = await Activity.BindAsync(dc, dc.State).ConfigureAwait(false);
            }

            string teamsChannelId = TeamsChannelId.GetValueOrNull(dc.State);
            if (string.IsNullOrEmpty(teamsChannelId))
            {
                teamsChannelId = dc.Context.Activity.TeamsGetChannelId();
            }

            Tuple<ConversationReference, string> result;

            if (dc.Context.Adapter is CloudAdapterBase)
            {
                // Retrieve the bot appid from TurnState's ClaimsIdentity
                string appId;
                if (dc.Context.TurnState.Get<ClaimsIdentity>(BotAdapter.BotIdentityKey) is ClaimsIdentity botIdentity)
                {
                    // Apparently 'version' is sometimes empty, which will result in no id returned from GetAppIdFromClaims
                    // TODO: Replace with new method
                    //appId = JwtTokenValidation.GetAppIdFromClaims(botIdentity.Claims);
                    appId = string.Empty;
                    if (string.IsNullOrEmpty(appId))
                    {
                        appId = botIdentity.Claims.FirstOrDefault(claim => claim.Type == AuthenticationConstants.AudienceClaim)?.Value;
                    }

                    if (string.IsNullOrEmpty(appId))
                    {
                        throw new InvalidOperationException($"Missing AppIdClaim in ClaimsIdentity.");
                    }
                }
                else
                {
                    throw new InvalidOperationException($"Missing {BotAdapter.BotIdentityKey} in {nameof(ITurnContext)} TurnState.");
                }

                // The result comes back as a tuple, which is used to set the two properties (if present).
                result = await TeamsInfo.SendMessageToTeamsChannelAsync(dc.Context, activity, teamsChannelId, appId, cancellationToken: cancellationToken).ConfigureAwait(false);
            }
            else
            {
                throw new InvalidOperationException($"The adapter does not support {nameof(SendMessageToTeamsChannel)}.");
            }

            if (ConversationReferenceProperty != null)
            {
                dc.State.SetValue(ConversationReferenceProperty.GetValue(dc.State), result.Item1);
            }
            
            if (ActivityIdProperty != null)
            {
                dc.State.SetValue(ActivityIdProperty.GetValue(dc.State), result.Item2);
            }

            return await dc.EndDialogAsync(result, cancellationToken: cancellationToken).ConfigureAwait(false);
        }

        /// <inheritdoc/>
        protected override string OnComputeId()
        {
            return $"{GetType().Name}[{TeamsChannelId?.ToString() ?? string.Empty},{ActivityIdProperty?.ToString() ?? string.Empty},{ConversationReferenceProperty?.ToString() ?? string.Empty}]";
        }
    }
}
